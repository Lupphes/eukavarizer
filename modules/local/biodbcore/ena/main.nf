process BIODBCORE_ENA {
    tag "$taxonomy_id"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "docker.io/luppo/biodbcore:1b8dd90e307a6937b09b9d7088f3b8ec5e8dacc8"

    input:
    val taxonomy_id
    val genome_size_ungapped
    val outdir
    val library_strategy
    val instrument_platform
    val minimum_coverage
    val maximum_coverage
    val max_results
    val assembly_quality

    output:
    path "*_ena_results.json", emit: ena_results
    path "*_samplesheet.csv"  , emit: samplesheet
    path "versions.yml"       , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${taxonomy_id}"
    def assembly_quality_arg = assembly_quality ? "--assembly_quality ${assembly_quality}" : ''
    """
    echo "No local sequencing data found. Downloading sequencing reads for taxonomy ID: ${taxonomy_id}"

    biodbcore \\
        --mode ena \\
        --taxonomy_id ${taxonomy_id} \\
        --library_strategy ${library_strategy} \\
        --instrument_platform ${instrument_platform} \\
        --minimum_coverage ${minimum_coverage} \\
        --maximum_coverage ${maximum_coverage} \\
        --max_results ${max_results} \\
        ${assembly_quality_arg} \\
        --outdir . \\
        --genome_size_ungapped ${genome_size_ungapped} \\
        ${args}

    # Rename outputs with configurable prefix
    mv ena_results.json ${prefix}_ena_results.json
    mv samplesheet.csv ${prefix}_samplesheet.csv

    echo "Sequencing data retrieval complete."
    cat ${prefix}_ena_results.json

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        biodbcore: \$( biodbcore --version 2>&1 | sed 's/biodbcore //g' )
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${taxonomy_id}"
    """
    touch ${prefix}_ena_results.json
    touch ${prefix}_samplesheet.csv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        biodbcore: \$( biodbcore --version 2>&1 | sed 's/biodbcore //g' )
    END_VERSIONS
    """
}
