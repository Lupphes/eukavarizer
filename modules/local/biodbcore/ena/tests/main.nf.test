nextflow_process {

    name "Test Process BIODBCORE_ENA"
    script "../main.nf"
    process "BIODBCORE_ENA"
    tag "biodbcore"
    tag "biodbcore/ena"
    tag "modules"
    tag "modules_local"

    test("Retrieve ENA data - yeast - stub") {

        options '-stub'

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = 4932  // Saccharomyces cerevisiae
                input[1] = 12000000  // genome_size_ungapped
                input[2] = "$outputDir"
                input[3] = "WGS"  // library_strategy
                input[4] = "ILLUMINA"  // instrument_platform
                input[5] = 10  // minimum_coverage
                input[6] = 100  // maximum_coverage
                input[7] = 5  // max_results
                input[8] = ""  // assembly_quality
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.ena_results },
                { assert process.out.samplesheet },
                { assert process.out.versions }
            )
        }
    }

    test("Retrieve ENA data with quality filter - stub") {

        options '-stub'

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = 511145  // E. coli K-12
                input[1] = 4600000  // genome_size_ungapped
                input[2] = "$outputDir"
                input[3] = "WGS"
                input[4] = "ILLUMINA"
                input[5] = 20
                input[6] = 200
                input[7] = 3
                input[8] = "high"
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.ena_results },
                { assert process.out.samplesheet },
                { assert process.out.versions }
            )
        }
    }

    test("Retrieve nanopore data - stub") {

        options '-stub'

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = 4932
                input[1] = 12000000
                input[2] = "$outputDir"
                input[3] = "WGS"
                input[4] = "OXFORD_NANOPORE"
                input[5] = 10
                input[6] = 100
                input[7] = 2
                input[8] = ""
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.ena_results },
                { assert process.out.samplesheet },
                { assert process.out.versions }
            )
        }
    }
}
