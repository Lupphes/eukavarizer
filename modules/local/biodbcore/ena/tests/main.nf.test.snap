{
    "Retrieve ENA data - yeast - stub": {
        "content": [
            {
                "0": [
                    "4932_ena_results.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "1": [
                    "4932_samplesheet.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "2": [
                    "versions.yml:md5,952be6a4eef859dbe2e61621daa92e17"
                ],
                "ena_results": [
                    "4932_ena_results.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "samplesheet": [
                    "4932_samplesheet.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,952be6a4eef859dbe2e61621daa92e17"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-20T22:34:26.50433574"
    },
    "Retrieve ENA data with quality filter - stub": {
        "content": [
            {
                "0": [
                    "511145_ena_results.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "1": [
                    "511145_samplesheet.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "2": [
                    "versions.yml:md5,952be6a4eef859dbe2e61621daa92e17"
                ],
                "ena_results": [
                    "511145_ena_results.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "samplesheet": [
                    "511145_samplesheet.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,952be6a4eef859dbe2e61621daa92e17"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-20T22:34:35.026391177"
    }
}